db-engine/src/main/java/com/splicemachine/db/impl/sql/catalog/DataDictionaryCache.java

```diff
@@ -359,19 +359,35 @@ public void clearStoredPreparedStatementCache() {
         storedPreparedStatementCache.invalidateAll();
     }
 
+    private long getDriverTaskTxnId() {
+        LanguageConnectionContext lcc =
+            (LanguageConnectionContext)ContextService.getCurrentContextManager().
+                                getContext(LanguageConnectionContext.CONTEXT_ID);
+        if (lcc == null)
+            return -1;
+        return lcc.getActiveStateTxId();
+    }
+
+    private Conglomerate txnAwareConglomerateCacheFind(Long conglomId) throws StandardException {
+        if (dd.useTxnAwareCache()) {
+            long txnId = getDriverTaskTxnId();
+            if (txnId == -1)
+                return null;
+            return txnAwareConglomerateCacheFind(new Pair(txnId,conglomId));
+        }
+        return null;
+    }
+
     public Conglomerate conglomerateCacheFind(TransactionController xactMgr,Long conglomId) throws StandardException {
         if (!dd.canReadCache(xactMgr) && conglomId>=DataDictionary.FIRST_USER_TABLE_NUMBER) {
             // Use cache even if dd says we can't as long as it's a system table (conglomID is < FIRST_USER_TABLE_NUMBER)
```

Comment:
I didn't add or change this comment, but I believe it's correct.  If conglomId < DataDictionary.FIRST_USER_TABLE_NUMBER then we don't enter this code, and we proceed to the line that reads from the data dictionary cache:
        return conglomerateCache.getIfPresent(conglomId);

It's confusing perhaps because the comment pertains to the call to conglomerateCache, not the txnAwareConglomerateCacheFind, which is right next to the comment (the comment existed before Yi added txnAwareConglomerateCacheFind).

So, it sounds like on upgrade, we could end up doing more dictionary lookups because of higher conglomerate numbers.  When I saw this code initially I did not understand how it's safe to always read system table conglomerates from cache.  I guess the code is depending on the mechanism which flushes items out when DDL is issued.  But if there is a race condition, could we perhaps ever have a stale version of a conglomerate descriptor in hand?  I don't know this area well enough, but I hope it's not something that could lead to dictionary corruption.  I don't see any isValid() method in Conglomerate like SPSDescriptor has.

Reply:
I see, yes, I was confused by the comment placement and the local cache, thanks for the explanation!

Yes, the idea is that it should be fine to use the cached descriptors if there are no DDL in-flight, but I'm not completely convinced this is right in 100% of the cases